.ci/test_r_package_windows.ps1

```diff
@@ -64,45 +67,67 @@ if ($env:COMPILER -eq "MINGW") {
-Rscript build_r.R --skip-install ; Check-Output $?
 
-$PKG_FILE_NAME = Get-Item *.tar.gz
-$LOG_FILE_NAME = "lightgbm.Rcheck/00check.log"
+# R CMD check is not used for MSVC builds
+if ($env:COMPILER -ne "MSVC") {
+  Rscript build_r.R --skip-install ; Check-Output $?
 
-$env:_R_CHECK_FORCE_SUGGESTS_ = 0
-if ($env:COMPILER -ne "MINGW") {
-  Write-Output "Running R CMD check without checking documentation"
-  R.exe CMD check --no-multiarch --no-examples --no-manual --ignore-vignettes ${PKG_FILE_NAME} ; $check_succeeded = $?
-} else {
+  $PKG_FILE_NAME = Get-Item *.tar.gz
+  $LOG_FILE_NAME = "lightgbm.Rcheck/00check.log"
+
+  $env:_R_CHECK_FORCE_SUGGESTS_ = 0
   Write-Output "Running R CMD check as CRAN"
   R.exe CMD check --no-multiarch --as-cran ${PKG_FILE_NAME} ; $check_succeeded = $?
-}
 
-Write-Output "R CMD check build logs:"
-Get-Content -Path $env:BUILD_SOURCESDIRECTORY\lightgbm.Rcheck\00install.out
-
-Check-Output $check_succeeded
+  Write-Output "R CMD check build logs:"
+  $INSTALL_LOG_FILE_NAME = "$env:BUILD_SOURCESDIRECTORY\lightgbm.Rcheck\00install.out"
+  Get-Content -Path "$INSTALL_LOG_FILE_NAME"
+
+  Check-Output $check_succeeded
+
+  Write-Output "Looking for issues with R CMD check results"
+  if (Get-Content "$LOG_FILE_NAME" | Select-String -Pattern "WARNING" -Quiet) {
+      echo "WARNINGS have been found by R CMD check!"
+      Check-Output $False
+  }
+
+  $note_str = Get-Content "${LOG_FILE_NAME}" | Select-String -Pattern ' NOTE' | Out-String ; Check-Output $?
+  $relevant_line = $note_str -match '.*Status: (\d+) NOTE.*'
+  $NUM_CHECK_NOTES = $matches[1]
+  $ALLOWED_CHECK_NOTES = 3
+  if ([int]$NUM_CHECK_NOTES -gt $ALLOWED_CHECK_NOTES) {
+      Write-Output "Found ${NUM_CHECK_NOTES} NOTEs from R CMD check. Only ${ALLOWED_CHECK_NOTES} are allowed"
+      Check-Output $False
+  }
+} else {
+  $INSTALL_LOG_FILE_NAME = "$env:BUILD_SOURCESDIRECTORY\00install_out.txt"
+  Rscript build_r.R *> $INSTALL_LOG_FILE_NAME ; $install_succeeded = $?
+  Write-Output "----- build and install logs -----"
+  Get-Content -Path "$INSTALL_LOG_FILE_NAME"
+  Write-Output "----- end of build and install logs -----"
+  Check-Output $install_succeeded
+  Write-Output "Running tests with testthat.R"
+  cd R-package/tests
+  Rscript testthat.R ; Check-Output $?
+}
 
-Write-Output "Looking for issues with R CMD check results"
-if (Get-Content "$LOG_FILE_NAME" | Select-String -Pattern "WARNING" -Quiet) {
-    echo "WARNINGS have been found by R CMD check!"
-    Check-Output $False
+# Checking that we actually got the expected compiler. The R package has some logic
+# to fail back to MinGW if MSVC fails, but for CI builds we need to check that the correct
+# compiler was used.
+if ($env:COMPILER -eq "MSVC") {
+  $checks = Select-String -Path "${INSTALL_LOG_FILE_NAME}" -Pattern 'Check for working CXX compiler.*MSVC'
+} else {
+  $checks = Select-String -Path "${INSTALL_LOG_FILE_NAME}" -Pattern 'Check for working CXX compiler.*mingw'
```

Comment:
I think it should be checked earlier right after printing these logs (before running `Rscript testthat.R`) to save CI time.

Reply:
ok sure, I can do that

Reply:
changed in https://github.com/microsoft/LightGBM/pull/2965/commits/828023f8c04002e00bfe66e43069bb383e7e0fea.

I didn't want to copy this code into both branches of the `if`, so my solution for your request was to move `testthat.R` further down the script and guard it with `if ($env:COMPILER -eq "MSVC")...`